### ISSUE TITLE
Simple snapshot export

### ISSUE THREAD

**Issue by der** [UKGovLD/registry-core#18]
- complementary to the “backup” function (which provides compressed n-quads)
- include only the specific versions of the registers and entities in the export - optionally including a cascade of sub-registers
- default behaviour to download a snapshot of the current resources in the subtree of the target register
- Initially would omit metadata (items) and versioning. Though this should be reviewed based on future ambition to support replication and round-trip editing.
- Export format to be discussed. JSON-LD has been suggested for CSV is an easier fit for tool chains. The W3C activity on data on the web might be relevant but requirements and timescales may ... diverge from our needs here.

Relates to  UKGovLD/ukl-registry-poc#82

**Comment by der** [comment]
As part of this have added functionality for CSV export of single items or whole registers, designed to support round tripping.

Import of such CSVs and the tree export (possibly via JSON-LD) TBD.

**Comment by 6a6d74** [comment]
@der - regarding choice of CSV or JSON-LD, please note (today!) the release of new WD from the W3C CSV on the Web WG that "the group considers to be feature complete and implementable".

The WD's are:

* [Model for Tabular Data and Metadata on the Web][model] - an abstract model for tabular data, and how to locate metadata that enables users to better understand what the data holds; this specification also contains non-normative guidance on how to parse CSV files
* [Metadata Vocabulary for Tabular Data][metadata] - a JSON-based format for expressing metadata about tabular data to inform validation, conversion, display and data entry for tabular data
* [Generating JSON from Tabular Data on the Web][csv2json] - how to convert tabular data into JSON
* [Generating RDF from Tabular Data on the Web][csv2rdf] - how to convert tabular data into RDF

[model]:http://www.w3.org/TR/2015/WD-tabular-data-model-20150416/
[metadata]:http://www.w3.org/TR/2015/WD-tabular-metadata-20150416/
[csv2json]:http://www.w3.org/TR/2015/WD-csv2json-20150416/
[csv2rdf]:http://www.w3.org/TR/2015/WD-csv2rdf-20150416/

CSV2RDF would be the path to get JSON-LD ... CSV2JSON is a simpler conversion "for those who need not care about RDF concepts".

The CSV2* docs both have fully featured examples ... 

Next steps for the WG are review and gathering more implementations for test against the [test suite][tests].

[tests]:http://w3c.github.io/csvw/tests/

**Comment by der** [comment]
I had looked at the earlier WD and felt they weren't suitable for this use case. In particular they focus on CSV2x rather than x2CSV and the mapping seems rather fixed.

My primary goal for the single register export/import is to enable round trip bulk editing so the CSV needs to be simple but the correspondence between the CSV and the register RDF is non trivial (needs to include status, and potentially other RegisterItem metadata, in the CSV alongside the entity properties).

**Comment by 6a6d74** [comment]
I think you should now find the CSV2* mapping more flexible - allowing for the publication of multiple resources per row in the CSV.

Regarding CSV2* vs. *2CSV, it's true that the spec is designed on the premise of working in one direction (starting from a CSV file). That said, CSV + metadata is _isomorphic_ with RDF, so the registry _export_ (RDF to CSV) would simply (!) need to publish a metadata file (in JSON) alongside the CSV.

My intervention here is a "by the way, did you know" rather than "we must use this solution" :-)

**Comment by der** [comment]
After some exploration I realize we have (at least) two different use case, needing different solutions and have implemented reasonably simple versions of both.

The two uses cases are *batch edit* and *state preservation*.

The *batch edit* case is where we want to make bulk changes to a register or set of registers. A key requirement here is that this be possible by domain exports without deep RDF knowledge. This is now supported by allowing CSV as a download and upload format and by providing an [edit](https://github.com/UKGovLD/registry-core/wiki/Api#batch-update--edit) operation which supports bulk patching. The operation is called *patch* in the UI because edit is already taken for interactive editing. So a typical workflow would be to download a register as a CSV, edit the CSV (possibly including adding new entries) and apply this as an edit/patch.

For the [CSV format](https://github.com/UKGovLD/registry-core/wiki/Api#csv-format) I've ended up with something which should be fairly readable and editable for simple cases but with brute force fall-backs for handling custom data types, language tags and bnode trees. I don't think this can be done easily using the CSV-on-Web machinery - that assumes uniform columns for things like typing and languages which would either mean imposing a schema on registry contents or would require dynamic generation of typed columns which would be complex for both users and implementors.

The state preservation use case is needed for selective backup/restore of the state of a register (tree) and to replicate a register tree from a master registry to slave registries. This use case in general requires history preservation not simply snapshot of latest values. However, it doesn't need to allow for non-expert editing of the state between backup and restore. 

This is now implemented by a pair of [import/export](https://github.com/UKGovLD/registry-core/wiki/Api#advanced-import-export) operations which deliver and consume an n-quads representation of the entire internal state for the selected register and its (recursive) contents.
